File: js/src/llm/fixtures.js
/**
 * fixtures.js — Fixture loader and MockTransport for tests.
 */

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { join, dirname } from 'node:path';

import { LLMError } from './errors.js';

const _thisDir = dirname(fileURLToPath(import.meta.url));
const FIXTURES_DIR = join(_thisDir, '..', '..', '..', 'tests', 'fixtures', 'llm');

/**
 * Load a named fixture JSON file.
 * @param {string} name  e.g. "complete_basic"
 * @returns {object}
 */
export function loadFixture(name) {
    const path = join(FIXTURES_DIR, `${name}.json`);
    try {
        const raw = readFileSync(path, 'utf-8');
        return JSON.parse(raw);
    } catch (err) {
        throw new Error(`fixture not found: ${path} (${err.message})`);
    }
}

/**
 * MockTransport — intercepts HTTP calls for testing.
 *
 * It records the last request (method, url, body) and returns the
 * canned response from the fixture.
 *
 * The transport interface matches what OllamaProvider._transport expects:
 *   request(method, url, body, timeoutS) → { status, body }
 *
 * For streaming fixtures (array response body), the body is returned as-is
 * (an array); OllamaProvider.chatStream handles that.
 */
export class MockTransport {
    /**
     * @param {object} fixture  Result of loadFixture()
     */
    constructor(fixture) {
        this._fixture = fixture;
        /** @type {{method:string, url:string, body:object|null}|null} */
        this.lastRequest = null;
    }

    /**
     * @param {string} method
     * @param {string} url
     * @param {object|null} body
     * @param {number} _timeoutS
     * @returns {{status: number, body: any}}
     */
    request(method, url, body, _timeoutS) {
        this.lastRequest = { method, url, body };
        const resp = this._fixture.response;
        return { status: resp.status, body: resp.body };
    }
}

/**
 * A MockTransport that always returns a 503 HTTP error.
 * Used for error tests.
 */
export class ErrorTransport {
    /**
     * @param {number} [status=503]
     */
    constructor(status = 503) {
        this._status = status;
    }

    request(_method, _url, _body, _timeoutS) {
        return { status: this._status, body: { error: 'service unavailable' } };
    }
}

/**
 * A MockTransport that simulates a connection refusal (network unreachable).
 * Used for error tests.
 */
export class UnreachableTransport {
    request(_method, _url, _body, _timeoutS) {
        throw new LLMError('unreachable', 'connect ECONNREFUSED 127.0.0.1:11434');
    }
}

/**
 * A MockTransport that simulates a timeout.
 * Used for error tests.
 */
export class TimeoutTransport {
    request(_method, _url, _body, _timeoutS) {
        throw new LLMError('timeout', 'Request timed out');
    }
}

/**
 * A MockTransport that returns a malformed JSON response for sentiment decode error test.
 */
export class MalformedJsonTransport {
    request(_method, _url, _body, _timeoutS) {
        // Return a valid HTTP 200 but the message content has no JSON object
        return {
            status: 200,
            body: {
                model: 'llama3.1:8b',
                message: { role: 'assistant', content: 'this is not json at all' },
                done: true,
                done_reason: 'stop',
                prompt_eval_count: 10,
                eval_count: 5,
                total_duration: 123456789,
            },
        };
    }
}

File: python/algoforge/llm/fixtures.py
from __future__ import annotations

import json
from pathlib import Path

from algoforge.llm.provider import LLMError

_FIXTURES_DIR = Path(__file__).parent.parent.parent.parent / "tests" / "fixtures" / "llm"


def load_fixture(name: str) -> dict:
    path = _FIXTURES_DIR / f"{name}.json"
    if not path.exists():
        raise FileNotFoundError(f"fixture not found: {path}")
    with path.open("r", encoding="utf-8") as fh:
        return json.load(fh)


class MockTransport:
    def __init__(self, fixture: dict) -> None:
        self._fixture = fixture
        self.last_request: dict | None = None

    def request(
        self,
        method: str,
        url: str,
        body: bytes | None,
        headers: dict[str, str],
        timeout: float,
    ) -> tuple[int, bytes]:
        parsed_body: dict | None = None
        if body is not None:
            parsed_body = json.loads(body)
        self.last_request = {
            "method": method,
            "url": url,
            "body": parsed_body,
        }
        resp = self._fixture["response"]
        status: int = resp["status"]
        resp_body = resp["body"]
        if isinstance(resp_body, list):
            raw = b"\n".join(json.dumps(chunk).encode("utf-8") for chunk in resp_body)
        else:
            raw = json.dumps(resp_body).encode("utf-8")
        return status, raw
